File: rock_paper_scissors.py
import random

options = []
default_game = False
name = input('Enter your name: ')
print(f'Hello, {name}')
score = 0
old_score = 0
new_usr = True
new_rating = []

with open('rating.txt', 'r') as f:
    for line in f:
        line_str = line.strip()
        line_list = line_str.split()
        if name == line_list[0]:
            new_usr = False
            score = int(line_list[1])
            old_score = int(line_list[1])


def new_user():
    with open('rating.txt', 'a') as o_file:
        o_file.write(f'{name} {score}\n')


def update_user():
    with open('rating.txt', 'r') as fin:
        for data in fin:
            data_str = data.strip()
            data_list = data_str.split()
            print(new_rating)
            if name == data_list[0]:
                new_rating.append(f'{name} {score}')
            else:
                new_rating.append(data_str)
            new_r_str = '\n'.join(new_rating)
            with open('rating.txt', 'w') as fout:
                fout.write(new_r_str + '\n')


def help_option():
    if default_game:
        game_type = 'classic version'
    else:
        game_type = 'your optional version'
    print(f'Hi, {name}. You are playing {game_type} of the game')
    print('Your available options for the game are:')
    print(options)
    print()
    print('To play a game, type one of the options')
    print('The option beats the next half of options, but loses to the previous ones')
    print('If you win, you get +100 points. Draw: +50 points. Lose: +0 points')
    print()
    print('Type "!exit" to exit your game. Your score will be added/updated to "rating.txt"')
    print('Type "!rating" to get your current score')
    print('Type "!help" to get this message one more time')


gestures = input('Press Enter for a classic game or write options separated by a comma (rock,paper,scissors,lizard)')
if gestures == '':
    options = ['rock', 'scissors', 'paper']
    default_game = True
else:
    options = gestures.split(',')

print("Okay, let's start")

if default_game is True:
    while True:
        action = input()
        c_action = random.choice(options)
        if action == '!exit':
            print('Bye!')
            if new_usr:
                new_user()
            else:
                update_user()
            break
        if action == '!rating':
            print(f'Your rating: {score}')
            continue
        if action == '!help':
            help_option()
            continue
        if action == 'rock' or action == 'scissors' or action == 'paper':
            if action == c_action:
                print(f'There is a draw ({action}) (+50 points)')
                score += 50
            elif action == 'rock' and c_action == 'scissors':
                print('Well done. Computer chose scissors and failed (+100 points)')
                score += 100
            elif action == 'scissors' and c_action == 'paper':
                print('Well done. Computer chose paper and failed (+100 points)')
                score += 100
            elif action == 'paper' and c_action == 'rock':
                print('Well done. Computer chose rock and failed (+100 points)')
                score += 100
            elif action == 'rock' and c_action == 'paper':
                print('Sorry, but computer chose paper')
            elif action == 'scissors' and c_action == 'rock':
                print('Sorry, but computer chose rock')
            elif action == 'paper' and c_action == 'scissors':
                print('Sorry, but computer chose scissors')
        else:
            print('Invalid input. Type "!help" for help')
elif len(options) % 2 != 0:
    while True:
        action = input()
        c_action = random.choice(options)
        if action == '!exit':
            print('Bye!')
            if new_usr:
                new_user()
            else:
                update_user()
            break
        if action == '!rating':
            print(f'Your rating: {score}')
            continue
        if action == '!help':
            help_option()
            continue
        if action in options:
            half_list = int((len(options) - 1) / 2)
            action_id = options.index(action)
            c_action_id = options.index(c_action)
            if action == c_action:
                print(f'There is a draw ({action}) (+50 points)')
                score += 50
            elif action_id + half_list <= len(options):
                if c_action_id > action_id >= c_action_id - half_list:
                    print(f'Sorry, but computer chose {c_action}')
                else:
                    print(f'Well done. Computer chose {c_action} and failed (+100 points)')
                    score += 100
            else:
                if c_action_id < action_id <= c_action_id + half_list:
                    print(f'Well done. Computer chose {c_action} and failed (+100 points)')
                    score += 100
                else:
                    print(f'Sorry, but computer chose {c_action}')
        else:
            print('Invalid input. Type "!help" for help')
elif len(options) % 2 == 0:
    while True:
        action = input()
        c_action = random.choice(options)
        if action == '!exit':
            print('Bye!')
            if new_usr:
                new_user()
            else:
                update_user()
            break
        if action == '!rating':
            print(f'Your rating: {score}')
            continue
        if action == '!help':
            help_option()
            continue
        if action in options:
            half_list = int(len(options) / 2)
            action_id = options.index(action)
            c_action_id = options.index(c_action)
            if action == c_action or action_id == c_action_id + half_list or action_id == c_action_id - half_list:
                print(f'There is a draw ({c_action}) (+50 points)')
                score += 50
            elif action_id + half_list <= len(options):
                if c_action_id > action_id >= c_action_id - half_list:
                    print(f'Sorry, but computer chose {c_action}')
                else:
                    print(f'Well done. Computer chose {c_action} and failed (+100 points)')
                    score += 100
            else:
                if c_action_id < action_id <= c_action_id + half_list:
                    print(f'Well done.Tim Computer chose {c_action} and failed (+100 points)')
                    score += 100
                else:
                    print(f'Sorry, but computer chose {c_action}')
        else:
            print('Invalid input. Type "!help" for help')

File: rock_paper_scissors.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

var (
	options     []string
	defaultGame bool
	name        string
	score       int
	newRating   []string
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	fmt.Print("Enter your name: ")
	if scanner.Scan() {
		name = scanner.Text()
		fmt.Printf("Hello, %s\n", name)
	}

	file, err := os.Open("rating.txt")
	if err != nil {
		fmt.Println("File not found.")
	}
	defer file.Close()

	fileScanner := bufio.NewScanner(file)
	for fileScanner.Scan() {
		line := fileScanner.Text()
		lineList := strings.Split(line, " ")
		if name == lineList[0] {
			score, _ = strconv.Atoi(lineList[1])
		}
	}

	fmt.Println("Enter gestures (comma-separated, leave empty for default - rock,scissors,paper): ")
	if scanner.Scan() {
		gestures := scanner.Text()
		if gestures == "" {
			options = []string{"rock", "scissors", "paper"}
			defaultGame = true
		} else {
			options = strings.Split(gestures, ",")
		}
	}

	fmt.Println("Okay, let's start")

	if defaultGame || len(options)%2 != 0 {
		playGame(scanner)
	} else {
		playGame(scanner)
	}
}

func playGame(scanner *bufio.Scanner) {
	random := rand.New(rand.NewSource(time.Now().UnixNano()))

	for {
		fmt.Print("Enter your move: ")
		scanner.Scan()
		action := scanner.Text()
		cAction := options[random.Intn(len(options))]

		switch action {
		case "!exit":
			fmt.Println("Bye!")
			updateUser()
			return
		case "!rating":
			fmt.Println("Your rating: ", score)
			continue
		case "!help":
			helpOption()
			continue
		default:
			if contains(options, action) {
				halfList := (len(options) - 1) / 2
				actionId := index(options, action)
				cActionId := index(options, cAction)

				if action == cAction {
					fmt.Printf("There is a draw (%s) (+50 points)\n", action)
					score += 50
				} else if actionId+halfList <= len(options) {
					if cActionId > actionId && actionId >= cActionId-halfList {
						fmt.Printf("Sorry, but computer chose %s\n", cAction)
					} else {
						fmt.Printf("Well done. Computer chose %s and failed (+100 points)\n", cAction)
						score += 100
					}
				} else {
					if cActionId < actionId && actionId <= cActionId+halfList {
						fmt.Printf("Well done. Computer chose %s and failed (+100 points)\n", cAction)
						score += 100
					} else {
						fmt.Printf("Sorry, but computer chose %s\n", cAction)
					}
				}
			} else {
				fmt.Println("Invalid input. Type \"!help\" for help")
			}
		}
	}
}

func newUser() {
	file, err := os.OpenFile("rating.txt", os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
	if err != nil {
		fmt.Println("An error occurred.")
	}
	defer file.Close()

	_, err = file.WriteString(name + " " + strconv.Itoa(score) + "\n")
	if err != nil {
		fmt.Println("An error occurred.")
	}
}

func updateUser() {
	file, err := os.Open("rating.txt")
	if err != nil {
		fmt.Println("An error occurred.")
	}
	defer file.Close()

	fileScanner := bufio.NewScanner(file)
	for fileScanner.Scan() {
		data := fileScanner.Text()
		dataList := strings.Split(data, " ")
		if name == dataList[0] {
			newRating = append(newRating, name+" "+strconv.Itoa(score))
		} else {
			newRating = append(newRating, data)
		}
	}

	writer, err := os.Create("rating.txt")
	if err != nil {
		fmt.Println("An error occurred.")
	}
	defer writer.Close()

	for _, s := range newRating {
		writer.WriteString(s + "\n")
	}
}

func helpOption() {
	gameType := "your optional version"
	if defaultGame {
		gameType = "classic version"
	}

	fmt.Printf("Hi, %s. You are playing %s of the game\n", name, gameType)
	fmt.Println("Your available options for the game are:")
	fmt.Println(options)
	fmt.Println()
	fmt.Println("To play a game, type one of the options")
	fmt.Println("The option beats the next half of options, but loses to the previous ones")
	fmt.Println("If you win, you get +100 points. Draw: +50 points. Lose: +0 points")
	fmt.Println()
	fmt.Println("Type \"!exit\" to exit your game. Your score will be added/updated to \"rating.txt\"")
	fmt.Println("Type \"!rating\" to get your current score")
	fmt.Println("Type \"!help\" to get this message one more time")
}

func contains(s []string, e string) bool {
	for _, a := range s {
		if a == e {
			return true
		}
	}
	return false
}

func index(s []string, e string) int {
	for i, a := range s {
		if a == e {
			return i
		}
	}
	return -1
}
